assets/js/workflow-diagram/WorkflowDiagram.tsx

```diff
@@ -240,11 +245,21 @@ export default function WorkflowDiagram(props: WorkflowDiagramProps) {
       } else if (isManualLayout) {
         // if isManualLayout, then we use values from store instead
         newModel.nodes.forEach(n => {
-          if (isManualLayout && n.type !== 'placeholder') {
-            n.position = fixedPositions[n.id];
-          }
+          if (n.type !== 'placeholder') n.position = fixedPositions[n.id];
         });
         setModel(newModel);
+      } else if (newModel.nodes.some(n => !hasXY(n))) {
+        // fallback: nodes lack positions → run layout now
+        const viewBounds = {
+          width: workflowDiagramRef.current?.clientWidth ?? 0,
+          height: workflowDiagramRef.current?.clientHeight ?? 0,
+        };
+        layout(newModel, setModel, flow, viewBounds, {
+          duration: props.layoutDuration ?? LAYOUT_DURATION,
+          forceFit: props.forceFit,
+        }).then(positions => {
+          chartCache.current.positions = positions;
+        });
```

Comment:
This part resolved the issue but then the question is that, do we know at what point in time we're triggering this layout animation?
Do we have control?